File: paleo_workbench/ui/workstation/mapping_stage_bar.py
"""``MappingStageBar``：编图阶段切换控件（AppBar 附近的紧凑分段条）。

风格（V5 §34/§62/§63）：compact / professional / desktop workstation，
不是大卡片；不做整页彩色主题，阶段只经 active 指示 + 小徽标表达。
切换瞬时（组可见性增量），绝不重载工程/画布。
"""
from __future__ import annotations

from PySide6.QtCore import Qt, Signal
from PySide6.QtWidgets import (
    QComboBox,
    QFrame,
    QHBoxLayout,
    QLabel,
    QWidget,
)

from paleo_workbench.mapping_workspace.stages import STAGE_ORDER, MappingStage


def _badge_tone(badge: str) -> str:
    if "!" in badge:
        return "error"
    if "~" in badge:
        return "warn"
    if "✓" in badge:
        return "ok"
    return "info"


def _repolish(widget: QWidget) -> None:
    style = widget.style()
    style.unpolish(widget)
    style.polish(widget)
    for child in widget.findChildren(QWidget):
        style.unpolish(child)
        style.polish(child)


class _StageSegment(QFrame):
    """单个阶段：序号圆点 + 短名 + 状态徽标。"""

    clicked = Signal()

    def __init__(self, stage: MappingStage, parent: QWidget | None = None):
        super().__init__(parent)
        self.stage = stage
        self.setObjectName("MappingStageSegment")
        self.setCursor(Qt.CursorShape.PointingHandCursor)
        self.setFocusPolicy(Qt.FocusPolicy.TabFocus)
        self.setAttribute(Qt.WidgetAttribute.WA_Hover, True)
        self.setToolTip(stage.description)
        self.setAccessibleName(stage.label)
        self._active = False
        self._badge = ""

        row = QHBoxLayout(self)
        row.setContentsMargins(6, 3, 10, 3)
        row.setSpacing(6)

        self._index = QLabel(str(stage.order + 1), self)
        self._index.setObjectName("MappingStageIndex")
        self._index.setAlignment(Qt.AlignmentFlag.AlignCenter)
        self._index.setFixedSize(18, 18)
        row.addWidget(self._index, 0)

        self._name = QLabel(stage.short_label, self)
        self._name.setObjectName("MappingStageName")
        row.addWidget(self._name, 0)

        self._badge_label = QLabel("", self)
        self._badge_label.setObjectName("MappingStageBadge")
        self._badge_label.hide()
        row.addWidget(self._badge_label, 0)

        self._refresh()

    def text(self) -> str:
        parts = [self.stage.short_label]
        if self._badge:
            parts.append(self._badge)
        return "  ".join(parts)

    def isChecked(self) -> bool:
        return self._active

    def click(self) -> None:
        self.clicked.emit()

    def set_active(self, active: bool) -> None:
        self._active = bool(active)
        self._refresh()

    def set_badge(self, badge: str) -> None:
        self._badge = str(badge or "")
        self._refresh()

    def mouseReleaseEvent(self, event) -> None:
        if event.button() == Qt.MouseButton.LeftButton:
            self.clicked.emit()
        super().mouseReleaseEvent(event)

    def keyPressEvent(self, event) -> None:
        if event.key() in (Qt.Key.Key_Return, Qt.Key.Key_Enter, Qt.Key.Key_Space):
            self.clicked.emit()
            event.accept()
            return
        super().keyPressEvent(event)

    def _refresh(self) -> None:
        self.setProperty("active", self._active)
        self.setProperty("stageIndex", self.stage.order)
        if self._badge:
            self._badge_label.setText(self._badge)
            self._badge_label.setProperty("tone", _badge_tone(self._badge))
            self._badge_label.show()
        else:
            self._badge_label.hide()
            self._badge_label.setText("")
        _repolish(self)


class MappingStageBar(QFrame):
    """阶段切换条：层位选择 + [ 智能预测 ] — [ 约束/单因素 ] — [ 综合编图 ]。

    相图按层位进行；层位是工程预设或从数据发现的目录，下拉选择后写入
    ``project.stratigraphy.target_horizon``（不可手输）。
    """

    stage_requested = Signal(str)  # MappingStage.value
    horizon_requested = Signal(str)

    def __init__(self, parent: QWidget | None = None):
        super().__init__(parent)
        self.setObjectName("MappingStageBar")
        self._buttons: dict[MappingStage, _StageSegment] = {}
        self._tracks: list[QFrame] = []
        self._suppress_horizon = False
        self._last_committed_horizon: str | None = None

        row = QHBoxLayout(self)
        row.setContentsMargins(10, 4, 10, 4)
        row.setSpacing(8)

        self.horizon_label = QLabel("层位", self)
        self.horizon_label.setObjectName("MappingStageMetaLabel")
        row.addWidget(self.horizon_label)
        self.horizon_combo = QComboBox(self)
        self.horizon_combo.setObjectName("MappingHorizonCombo")
        self.horizon_combo.setEditable(False)
        self.horizon_combo.setInsertPolicy(QComboBox.InsertPolicy.NoInsert)
        self.horizon_combo.setFixedWidth(160)
        self.horizon_combo.setPlaceholderText("选择层位")
        # R2 P2-3：紧凑视口隐藏「层位」前缀标签后，选中态唯一的语境说明；
        # 屏幕阅读器需要显式名（组合框本身不读占位文本）。
        self.horizon_combo.setAccessibleName("层位")
        self.horizon_combo.setToolTip(
            "编图层位：从工程层序格架或导入的层位数据中选择（相图按层位进行）")
        self.horizon_combo.currentIndexChanged.connect(
            lambda _i: self._commit_horizon())
        row.addWidget(self.horizon_combo, 0)

        divider = QFrame(self)
        divider.setObjectName("MappingStageDivider")
        divider.setFixedSize(1, 18)
        row.addWidget(divider)

        for index, stage in enumerate(STAGE_ORDER):
            segment = _StageSegment(stage, self)
            segment.clicked.connect(
                lambda value=stage.value: self.stage_requested.emit(value))
            self._buttons[stage] = segment
            row.addWidget(segment, 0)
            if index < len(STAGE_ORDER) - 1:
                track = QFrame(self)
                track.setObjectName("MappingStageTrack")
                track.setFixedSize(28, 2)
                row.addWidget(track, 0, Qt.AlignmentFlag.AlignVCenter)
                self._tracks.append(track)
        row.addStretch(1)

    def set_viewport_class(self, viewport) -> None:
        """V9 viewport 策略：紧凑视口隐藏「层位」前缀标签（下拉自带占位
        文案，语义不丢失），换取阶段条横向空间。"""
        from paleo_workbench.ui.dock_framework import ViewportClass

        self.horizon_label.setVisible(viewport is not ViewportClass.COMPACT)

    # -- 状态同步（由宿主/MappingStageController 驱动） --------------------------

    def set_current_stage(self, stage_value: str) -> None:
        from paleo_workbench.mapping_workspace.stages import stage_from_value

        stage = stage_from_value(stage_value)
        for button_stage, button in self._buttons.items():
            button.set_active(stage is not None and button_stage == stage)
        current_order = stage.order if stage is not None else -1
        for index, track in enumerate(self._tracks):
            track.setProperty("complete", index < current_order)
            _repolish(track)

    def set_stage_badge(self, stage: MappingStage, badge: str) -> None:
        button = self._buttons.get(stage)
        if button is not None:
            button.set_badge(badge)

    def refresh_badges(self, badges: dict[str, str]) -> None:
        """批量更新徽标（key = stage.value）。"""
        from paleo_workbench.mapping_workspace.stages import stage_from_value

        for value, badge in (badges or {}).items():
            stage = stage_from_value(value)
            if stage is not None:
                self.set_stage_badge(stage, badge)

    def current_horizon(self) -> str:
        return self.horizon_combo.currentText().strip()

    def set_horizon_state(self, horizon: str, options: list[str] | None = None) -> None:
        """同步工程层位到选择框（不发 ``horizon_requested``）。"""
        target = str(horizon or "").strip()
        choices: list[str] = []
        for name in list(options or []):
            text = str(name or "").strip()
            if text and text not in choices:
                choices.append(text)
        if target and target not in choices:
            choices.insert(0, target)
        self._suppress_horizon = True
        self.horizon_combo.clear()
        if choices:
            self.horizon_combo.addItems(choices)
        if target:
            index = self.horizon_combo.findText(target)
            if index < 0:
                self.horizon_combo.insertItem(0, target)
                index = 0
            self.horizon_combo.setCurrentIndex(index)
        else:
            self.horizon_combo.setCurrentIndex(-1)
        self._last_committed_horizon = target or None
        self._suppress_horizon = False

    def _commit_horizon(self) -> None:
        if self._suppress_horizon:
            return
        text = self.current_horizon()
        if text == self._last_committed_horizon:
            return
        self._last_committed_horizon = text or None
        self.horizon_requested.emit(text)

File: libs/ui_composite/src/mapping_stage_bar.cpp
#include <pwb/ui_composite/mapping_stage_bar.hpp>

#include <pwb/ui_widgets/ui_context.hpp>

#include <algorithm>

#include <QHBoxLayout>
#include <QKeyEvent>
#include <QMouseEvent>

namespace pwb::ui_composite {

namespace {

using pwb::tool_policy::MappingStage;

QString badge_tone(const QString& badge) {
    if (badge.contains('!')) return QStringLiteral("error");
    if (badge.contains('~')) return QStringLiteral("warn");
    if (badge.contains(QChar(0x2713))) return QStringLiteral("ok");  // ✓
    return QStringLiteral("info");
}

int stage_order(MappingStage stage) {
    for (size_t i = 0; i < pwb::tool_policy::kStageOrder.size(); ++i) {
        if (pwb::tool_policy::kStageOrder[i] == stage) {
            return static_cast<int>(i);
        }
    }
    return -1;
}

}  // namespace

StageSegment::StageSegment(MappingStage stage, QWidget* parent)
    : QFrame(parent), stage(stage) {
    setObjectName("MappingStageSegment");
    setCursor(Qt::PointingHandCursor);
    setFocusPolicy(Qt::TabFocus);
    setAttribute(Qt::WA_Hover, true);
    setToolTip(QString::fromUtf8(
        pwb::tool_policy::stage_description(stage)));
    setAccessibleName(QString::fromUtf8(
        pwb::tool_policy::stage_label(stage)));

    auto* row = new QHBoxLayout(this);
    row->setContentsMargins(4, 2, 6, 2);
    row->setSpacing(6);

    index_ = new QLabel(QString::number(stage_order(stage) + 1), this);
    index_->setObjectName("MappingStageIndex");
    index_->setAlignment(Qt::AlignCenter);
    index_->setFixedSize(18, 18);
    row->addWidget(index_, 0);

    name_ = new QLabel(
        QString::fromUtf8(pwb::tool_policy::stage_short_label(stage)),
        this);
    name_->setObjectName("MappingStageName");
    row->addWidget(name_, 0);

    badge_label_ = new QLabel(QString(), this);
    badge_label_->setObjectName("MappingStageBadge");
    badge_label_->hide();
    row->addWidget(badge_label_, 0);

    refresh();
}

QString StageSegment::text() const {
    QString out = name_->text();
    if (!badge_.isEmpty()) out += QStringLiteral("  ") + badge_;
    return out;
}

void StageSegment::set_active(bool active) {
    active_ = active;
    refresh();
}

void StageSegment::set_badge(const QString& badge) {
    badge_ = badge;
    refresh();
}

void StageSegment::mouseReleaseEvent(QMouseEvent* event) {
    if (event->button() == Qt::LeftButton) emit clicked();
    QFrame::mouseReleaseEvent(event);
}

void StageSegment::keyPressEvent(QKeyEvent* event) {
    if (event->key() == Qt::Key_Return || event->key() == Qt::Key_Enter ||
        event->key() == Qt::Key_Space) {
        emit clicked();
        event->accept();
        return;
    }
    QFrame::keyPressEvent(event);
}

void StageSegment::refresh() {
    setProperty("active", active_);
    setProperty("stageIndex", stage_order(stage));
    if (!badge_.isEmpty()) {
        badge_label_->setText(badge_);
        badge_label_->setProperty("tone", badge_tone(badge_));
        badge_label_->show();
    } else {
        badge_label_->hide();
        badge_label_->setText(QString());
    }
    pwb::ui_widgets::repolish(this);
}

// ---------------------------------------------------------------------------

MappingStageBar::MappingStageBar(QWidget* parent) : QFrame(parent) {
    setObjectName("MappingStageBar");

    auto* row = new QHBoxLayout(this);
    row->setContentsMargins(8, 2, 8, 2);
    row->setSpacing(6);

    horizon_label = new QLabel(QStringLiteral("层位"), this);
    horizon_label->setObjectName("MappingStageMetaLabel");
    row->addWidget(horizon_label);
    horizon_combo = new QComboBox(this);
    horizon_combo->setObjectName("MappingHorizonCombo");
    horizon_combo->setEditable(false);
    horizon_combo->setInsertPolicy(QComboBox::NoInsert);
    horizon_combo->setMinimumWidth(130);
    horizon_combo->setMaximumWidth(130);
    horizon_combo->setPlaceholderText(QStringLiteral("选择层位"));
    horizon_combo->setAccessibleName(QStringLiteral("层位"));
    horizon_combo->setToolTip(QStringLiteral(
        "编图层位：从工程层序格架或导入的层位数据中选择（相图按层位进行）"));
    connect(horizon_combo, &QComboBox::currentIndexChanged, this,
            [this](int) { commit_horizon(); });
    row->addWidget(horizon_combo, 0);

    auto* divider = new QFrame(this);
    divider->setObjectName("MappingStageDivider");
    divider->setFixedSize(1, 18);
    row->addWidget(divider);

    const auto& order = pwb::tool_policy::kStageOrder;
    for (size_t index = 0; index < order.size(); ++index) {
        const MappingStage stage = order[index];
        auto* segment = new StageSegment(stage, this);
        connect(segment, &StageSegment::clicked, this,
                [this, stage]() {
                    emit stage_requested(QString::fromUtf8(
                        pwb::tool_policy::stage_value(stage)));
                });
        buttons_[stage] = segment;
        row->addWidget(segment, 0);
        if (index + 1 < order.size()) {
            auto* track = new QFrame(this);
            track->setObjectName("MappingStageTrack");
            track->setFixedSize(20, 2);
            row->addWidget(track, 0, Qt::AlignVCenter);
            tracks_.push_back(track);
        }
    }
    // 无尾 stretch：阶段条 sizeHint 必须等于内容宽（V7 同行契约）。
}

void MappingStageBar::set_current_stage(const std::string& stage_value) {
    const auto stage = pwb::tool_policy::stage_from_value(stage_value);
    for (auto& [button_stage, button] : buttons_) {
        button->set_active(stage.has_value() && button_stage == *stage);
    }
    const int current_order =
        stage.has_value() ? stage_order(*stage) : -1;
    for (size_t index = 0; index < tracks_.size(); ++index) {
        tracks_[index]->setProperty("complete",
                                    static_cast<int>(index) < current_order);
        pwb::ui_widgets::repolish(tracks_[index]);
    }
}

void MappingStageBar::set_stage_badge(MappingStage stage,
                                      const QString& badge) {
    auto it = buttons_.find(stage);
    if (it != buttons_.end()) it->second->set_badge(badge);
}

void MappingStageBar::refresh_badges(
    const std::map<std::string, QString>& badges) {
    for (const auto& [value, badge] : badges) {
        const auto stage = pwb::tool_policy::stage_from_value(value);
        if (stage.has_value()) set_stage_badge(*stage, badge);
    }
}

QString MappingStageBar::current_horizon() const {
    return horizon_combo->currentText().trimmed();
}

void MappingStageBar::set_horizon_state(
    const QString& horizon, const std::vector<QString>& options) {
    const QString target = horizon.trimmed();
    std::vector<QString> choices;
    for (const QString& name : options) {
        const QString text = name.trimmed();
        if (!text.isEmpty() &&
            std::find(choices.begin(), choices.end(), text) ==
                choices.end()) {
            choices.push_back(text);
        }
    }
    if (!target.isEmpty() &&
        std::find(choices.begin(), choices.end(), target) ==
            choices.end()) {
        choices.insert(choices.begin(), target);
    }
    suppress_horizon_ = true;
    horizon_combo->clear();
    for (const QString& choice : choices) horizon_combo->addItem(choice);
    if (!target.isEmpty()) {
        int index = horizon_combo->findText(target);
        if (index < 0) {
            horizon_combo->insertItem(0, target);
            index = 0;
        }
        horizon_combo->setCurrentIndex(index);
    } else {
        horizon_combo->setCurrentIndex(-1);
    }
    last_committed_horizon_ = target;
    suppress_horizon_ = false;
}

void MappingStageBar::commit_horizon() {
    if (suppress_horizon_) return;
    const QString text = current_horizon();
    if (text == last_committed_horizon_) return;
    last_committed_horizon_ = text;
    emit horizon_requested(text);
}

}  // namespace pwb::ui_composite
